fri/src/prover.rs

```diff
@@ -1,126 +1,111 @@
 
-    let pow_witness = challenger.grind(config.proof_of_work_bits);
+    let commit_phase_result = commit_phase(g, config, inputs, challenger);
 
-    let query_indices: Vec<usize> = (0..config.num_queries)
-        .map(|_| challenger.sample_bits(log_max_height))
-        .collect();
+    let pow_witness = challenger.grind(config.proof_of_work_bits);
 
     let query_proofs = info_span!("query phase").in_scope(|| {
-        query_indices
-            .iter()
-            .map(|&index| answer_query(config, &commit_phase_result.data, index))
+        iter::repeat_with(|| challenger.sample_bits(log_max_height + g.extra_query_index_bits()))
+            .take(config.num_queries)
+            .map(|index| QueryProof {
+                input_proof: open_input(index),
+                commit_phase_openings: answer_query(
+                    config,
+                    &commit_phase_result.data,
+                    index >> g.extra_query_index_bits(),
+                ),
+            })
             .collect()
     });
 
-    (
-        FriProof {
-            commit_phase_commits: commit_phase_result.commits,
-            query_proofs,
-            final_poly: commit_phase_result.final_poly,
-            pow_witness,
-        },
-        query_indices,
-    )
+    FriProof {
+        commit_phase_commits: commit_phase_result.commits,
+        query_proofs,
+        final_poly: commit_phase_result.final_poly,
+        pow_witness,
+    }
 }
 
-fn answer_query<F, M>(
-    config: &FriConfig<M>,
-    commit_phase_commits: &[M::ProverData<RowMajorMatrix<F>>],
-    index: usize,
-) -> QueryProof<F, M>
-where
-    F: Field,
-    M: Mmcs<F>,
-{
-    let commit_phase_openings = commit_phase_commits
-        .iter()
-        .enumerate()
-        .map(|(i, commit)| {
-            let index_i = index >> i;
-            let index_i_sibling = index_i ^ 1;
-            let index_pair = index_i >> 1;
-
-            let (mut opened_rows, opening_proof) = config.mmcs.open_batch(index_pair, commit);
-            assert_eq!(opened_rows.len(), 1);
-            let opened_row = opened_rows.pop().unwrap();
-            assert_eq!(opened_row.len(), 2, "Committed data should be in pairs");
-            let sibling_value = opened_row[index_i_sibling % 2];
-
-            CommitPhaseProofStep {
-                sibling_value,
-                opening_proof,
-            }
-        })
-        .collect();
-
-    QueryProof {
-        commit_phase_openings,
-    }
+struct CommitPhaseResult<F: Field, M: Mmcs<F>> {
+    commits: Vec<M::Commitment>,
+    data: Vec<M::ProverData<RowMajorMatrix<F>>>,
+    final_poly: F,
 }
 
 #[instrument(name = "commit phase", skip_all)]
-fn commit_phase<F, M, Challenger>(
+fn commit_phase<G, F, M, Challenger>(
+    g: &G,
     config: &FriConfig<M>,
-    input: &[Option<Vec<F>>; 32],
-    log_max_height: usize,
+    inputs: Vec<Vec<F>>,
     challenger: &mut Challenger,
 ) -> CommitPhaseResult<F, M>
 where
-    F: TwoAdicField,
+    F: Field,
```

Comment:
I understand that `F = Mersennes31` does not have large two-adic multiplicative subgroups, unless we commit over a complex extension of it. 

But, shouldn't the api require that we commit over at least a two-adic field extension ? I wonder, if we could mess up and try to use the api with 'non-smooth' fields (together with non-smooth extensions).

Reply:
This PR changes fri so it's generic over the folding function; so there's no bound required. two_adic_pcs still requires TwoAdicField